endpoints/openrtb2/auction.go

```diff
@@ -653,17 +654,40 @@ func mapSChains(req *openrtb_ext.RequestWrapper) error {
 	return nil
 }
 
-func validateAndFillSourceTID(req *openrtb2.BidRequest) error {
+func validateAndFillSourceTID(req *openrtb_ext.RequestWrapper, generateRequestID bool, hasStoredBidRequest bool, isAmp bool) error {
 	if req.Source == nil {
 		req.Source = &openrtb2.Source{}
 	}
-	if req.Source.TID == "" {
-		if rawUUID, err := uuid.NewV4(); err == nil {
-			req.Source.TID = rawUUID.String()
-		} else {
-			return errors.New("req.Source.TID missing in the req and error creating a random UID")
+
+	rawUUID, err := uuid.NewV4()
+	if err != nil {
+		return errors.New("error creating a random UUID for source.tid")
+	}
+
+	if req.Source.TID == "" || req.Source.TID == "{{UUID}}" {
+		req.Source.TID = rawUUID.String()
+	}
+
+	if generateRequestID && (isAmp || hasStoredBidRequest) {
```

Comment:
We may want to put this if statement in an `else` from the `if` above, so that we don't call `rawUUID.String()` twice.

Actually we probably want to put it all into one, so we don't call `uuid.NewV4()` when we don't have to, and potentially error out when we don't even have to generate a random UUID for the request.

```
	if  req.Source.TID == "" || req.Source.TID == "{{UUID}}" || ( generateRequestID && (isAmp || hasStoredBidRequest) ) {
	    rawUUID, err := uuid.NewV4()
	    if err != nil {
		return errors.New("error creating a random UUID for source.tid")
	    }
	    req.Source.TID = rawUUID.String()
        }
```

Reply:
thanks for the review, updated